### PR TITLE
Codemod: parameterization of file path in Flask's send-file

### PR DESCRIPTION
**Pull request by andrecsilva**
## Overview
* A new codemod that parameterizes paths given as input to Flask `send-file` function.
* Adds some functionality to generate available names in a given scope.

### CODE REVIEW

File: src/core_codemods/docs/pixee_python_flask-send-file-path-parameterization.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+The `send_file` function from Flask is susceptible to a path injection attack if its input is not proper validated. 
+In a path injection attack, the malicious agent can craft a path containing special paths like `./` or `../` to resolve an unintended file. This allows the agent to overwrite, delete or read arbitrary files.
```
Comment (drdavella) [inline src/core_codemods/docs/pixee_python_flask-send-file-path-parameterization.md]:
```suggestion
In a path traversal attack, the malicious agent can craft a path containing special paths like `./` or `../` to resolve a file outside of the expected directory path. This potentially allows the agent to overwrite, delete or read arbitrary files. In the case of `flask.send_file`, the result is that a malicious user could potentially download sensitive files that exist on the filesystem where the application is being hosted.
```

File: src/core_codemods/docs/pixee_python_flask-send-file-path-parameterization.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+The `send_file` function from Flask is susceptible to a path injection attack if its input is not proper validated.
```
Comment (drdavella) [inline src/core_codemods/docs/pixee_python_flask-send-file-path-parameterization.md]:
```suggestion
The `send_file` function from Flask is susceptible to a path traversal attack if its input is not proper validated. 
```

**Comment by clavedeluna** [comment]
Does sonarcloud provide in-file or settings to let it know we expect insecure code in samples?

File: src/core_codemods/flask_send_file_path_parameterization.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+from codemodder.codemods.api import BaseCodemod
+from codemodder.codemods.base_codemod import ReviewGuidance
+from codemodder.codemods.utils import BaseType, infer_expression_type
+from codemodder.codemods.utils_mixin import NameAndAncestorResolutionMixin
+from codemodder.utils.utils import positional_to_keyword
+
+
+class FlaskSendFilePathParameterization(BaseCodemod, NameAndAncestorResolutionMixin):
+    NAME = "flask-send-file-path-parameterization"
+    SUMMARY = "Replaces `send_file` with `send_from_directory`"
```
Comment (drdavella) [inline src/core_codemods/flask_send_file_path_parameterization.py]:
```suggestion
    SUMMARY = "Replace unsafe usage of `flask.send_file`"
```

File: src/core_codemods/docs/pixee_python_flask-send-file-path-parameterization.md (hunk 2 above)
Comment (clavedeluna) [inline src/core_codemods/docs/pixee_python_flask-send-file-path-parameterization.md]:
```suggestion
The `Flask` `send_file` function is susceptible to a path injection attack if its input is not properly validated. 
```

**Review by drdavella — APPROVED** [review]
This is a nice codemod. Just a few small requests for metadata updates before merging.

File: src/core_codemods/flask_send_file_path_parameterization.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+from typing import Optional
+from codemodder.codemods.api import BaseCodemod
+from codemodder.codemods.base_codemod import ReviewGuidance
+from codemodder.codemods.utils import BaseType, infer_expression_type
+from codemodder.codemods.utils_mixin import NameAndAncestorResolutionMixin
+from codemodder.utils.utils import positional_to_keyword
+
+
+class FlaskSendFilePathParameterization(BaseCodemod, NameAndAncestorResolutionMixin):
+    NAME = "flask-send-file-path-parameterization"
```
Comment (clavedeluna) [inline src/core_codemods/flask_send_file_path_parameterization.py]: I maybe would've called it `replace-flask-send-file` or something without the word parametrization, but maybe Dan has stronger opinions

File: src/core_codemods/flask_send_file_path_parameterization.py (hunk 3 above)
Comment (clavedeluna) [inline src/core_codemods/flask_send_file_path_parameterization.py]: I'd add `Flask` somewhere in the summary / description

File: tests/codemods/test_flask_send_file_path_parameterization.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
+from core_codemods.flask_send_file_path_parameterization import (
+    FlaskSendFilePathParameterization,
+)
+from tests.codemods.base_codemod_test import BaseCodemodTest
+from textwrap import dedent
```
Comment (clavedeluna) [inline tests/codemods/test_flask_send_file_path_parameterization.py]: can remove dedent use and rely on BaseCodemodTest doing it

File: tests/codemods/test_flask_send_file_path_parameterization.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+        from flask import Flask, send_file
+
+        app = Flask(__name__)
+
+        @app.route("/uploads/<path:name>")
+        def download_file(name):
+            return send_file(name)
+        """
+        self.run_and_assert(tmpdir, dedent(input_code), dedent(input_code))
+        assert len(self.file_context.codemod_changes) == 0
```
Comment (clavedeluna) [inline tests/codemods/test_flask_send_file_path_parameterization.py]: not seeing a basic param of just a string without interpolation, like `send_file('path')`

File: tests/samples/flask_send_file_path_parameterization.py [hunk 7]
```diff
@@ -0,0 +1,7 @@
+from flask import Flask, send_file
+
+app = Flask(__name__)
+
+@app.route("/uploads/<path:name>")
+def download_file(name):
+    return send_file(f'path/to/{name}.txt')
```
[the remark this answers is not in the document]
Comment (clavedeluna) [reply tests/samples/flask_send_file_path_parameterization.py]: ah I guess you can just dismiss it to say it's tests

**Review by clavedeluna — CHANGES_REQUESTED** [review]

File: src/core_codemods/flask_send_file_path_parameterization.py (hunk 4 above)
Comment (drdavella) [reply src/core_codemods/flask_send_file_path_parameterization.py]: I like your shorter suggestion.

Comment (andrecsilva) [reply src/core_codemods/flask_send_file_path_parameterization.py]: Next in my wishlist: a codemod that renames a codemod in `codemodder-python`.

**Review by clavedeluna — APPROVED** [review]

**Pull request merged**